cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/SourceControlManagementHttpHandler.java

```diff
@@ -104,6 +125,80 @@ public void deleteRepository(FullHttpRequest request, HttpResponder responder,
                                                               namespaceId));
   }
 
+  /**
+   * Pushes a set of applications configs to linked repository in Json format. It expects a post body that has an
+   * optional commit message and an array of objects, with each object specifying the application and version. The
+   * size of applications list should not be more than 10.
+   * E.g.
+   *
+   * <pre>
+   * {@code
+   * {
+   *   "apps": [
+   *   {"application":"XYZ", "version": "1.23"},
+   *   {"application":"ABC", "version": "1.2.4"},
+   *   {"application":"FOO", "version": "1.23.4"}
+   *   ],
+   *   "commitMessage": "pushed apps XYZ, ABC and FOO"
+   * }
+   * }
+   *
+   * </pre>
+   * The response will be an array of {@link PushAppResponse} object, which encapsulates the application name,
+   * version and fileHash.
+   */
+  @POST
+  @Path("/push")
+  public void pushApps(FullHttpRequest request, HttpResponder responder,
+                       @PathParam("namespace-id") String namespaceId) throws Exception {
+    checkSourceControlFeatureFlag();
+    NamespaceId namespace = validateNamespaceId(namespaceId);
+    ImmutablePair<List<ApplicationId>, String> appsRequest = validateAndGetAppsRequest(request, namespace);
+
+    try {
+      PushAppsResponse pushResponse = sourceControlService.pushApps(namespace, appsRequest.getFirst(),
+                                                                    appsRequest.getSecond());
+      responder.sendJson(HttpResponseStatus.OK, GSON.toJson(pushResponse));
+    } catch (NoChangesToPushException e) {
+      // TODO: CDAP-20383 define the case that the applications to push do not have any changes
+      responder.sendString(HttpResponseStatus.BAD_REQUEST, e.getMessage());
+    } catch (PushFailureException e) {
+      responder.sendString(HttpResponseStatus.INTERNAL_SERVER_ERROR, e.getMessage());
+    }
+  }
+
+  private ImmutablePair<List<ApplicationId>, String> validateAndGetAppsRequest(
+    FullHttpRequest request, NamespaceId namespace) throws BadRequestException {
+    
+    PushAppsRequest appsRequest;
+    try {
+      appsRequest = parseBody(request, PushAppsRequest.class);
+    } catch (JsonSyntaxException e) {
+      throw new BadRequestException("Invalid request body: " + e.getMessage());
+    }
+    
+    if (appsRequest == null || appsRequest.getApps() == null || appsRequest.getApps().isEmpty()) {
+      throw new BadRequestException("Please specify a list of applications.");
+    }
+
+    if (appsRequest.getApps().size() >
+      cConf.getInt(Constants.SourceControlManagement.GIT_REPOSITORIES_PUSH_APPS_COUNT_LIMIT)) {
+      throw new BadRequestException("Please push no more than 10 applications at one time.");
```

Comment:
Do you want to use the same constant in place of 10 in the error message?

Reply:
good catch, fixed to use the one from cConf